include/ensmallen_bits/adam/padam_update.hpp

```diff
@@ -0,0 +1,153 @@
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_ADAM_PADAM_UPDATE_HPP
+#define ENSMALLEN_ADAM_PADAM_UPDATE_HPP
+
+namespace ens {
+
+/**
+ * Partially adaptive momentum estimation method (Padam),
+ * adopts historical gradient information to automatically adjust the
+ * learning rate.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @article{
+ *   title   = {Closing the Generalization Gap of Adaptive Gradient Methods in
+ *              Training Deep Neural Networks},
+ *   author  = {{Chen}, J. and {Gu}, Q.},
+ *   journal = {ArXiv e-prints},
+ *   url     = {https://arxiv.org/abs/1806.06763}
+ *   year    = {2018}
+ * }
+ * @endcode
+ */
+class PadamUpdate
+{
+ public:
+  /**
+   * Construct the Padam update policy with the given parameters.
+   *
+   * @param epsilon The epsilon value used to initialise the squared gradient
+   *        parameter.
+   * @param beta1 The smoothing parameter.
+   * @param beta2 The second moment coefficient.
+   * @param partial Partially adaptive parameter.
+   */
+  PadamUpdate(const double epsilon = 1e-8,
+              const double beta1 = 0.9,
+              const double beta2 = 0.999,
+              const double partial = 0.25) :
+      epsilon(epsilon),
+      beta1(beta1),
+      beta2(beta2),
+      partial(partial),
+      iteration(0)
+  {
+    // Nothing to do.
+  }
+
+  /**
+   * The Initialize method is called by SGD Optimizer method before the start of
+   * the iteration update process.
+   *
+   * @param rows Number of rows in the gradient matrix.
+   * @param cols Number of columns in the gradient matrix.
+   */
+  void Initialize(const size_t rows, const size_t cols)
+  {
+    m = arma::zeros<arma::mat>(rows, cols);
+    v = arma::zeros<arma::mat>(rows, cols);
+    vImproved = arma::zeros<arma::mat>(rows, cols);
+  }
+
+  /**
+   * Update step for Padam.
+   *
+   * @param iterate Parameters that minimize the function.
+   * @param stepSize Step size to be used for the given iteration.
+   * @param gradient The gradient matrix.
+   */
+  void Update(arma::mat& iterate,
+              const double stepSize,
+              const arma::mat& gradient)
+  {
+    // Increment the iteration counter variable.
+    ++iteration;
+
+    // And update the iterate.
+    m *= beta1;
+    m += (1 - beta1) * gradient;
+
+    v *= beta2;
+    v += (1 - beta2) * (gradient % gradient);
+
+    const double biasCorrection1 = 1.0 - std::pow(beta1, iteration);
+    const double biasCorrection2 = 1.0 - std::pow(beta2, iteration);
+
+    // Element wise maximum of past and present squared gradients.
+    vImproved = arma::max(vImproved, v);
+
+    iterate -= (stepSize * std::sqrt(biasCorrection2) / biasCorrection1) *
+        m / arma::pow(arma::sqrt(vImproved) + epsilon, partial * 2);
```

Comment:
It's possible I have misunderstood something here, but my reading of Algorithm 1 in the paper suggests that this could be `m / arma::pow(vImproved + epsilon, partial)`.  But I think this depends on how the `epsilon` correction should be applied.  Personally I don't think it makes particularly much difference whether it is applied to the square root of `vImproved` or not (the suggestion I wrote would be a little quicker in practice).  But maybe there is some reason I don't know of to write it the way you did?

(In the end I'm indifferent on how it gets implemented, mostly I am just curious here.)